extends Area2D

@export var speed := 300.0
var direction := Vector2.RIGHT

@export var damage := 1
const EXPLOSION = preload("res://effects/explosion_effect.tscn")

# Called when the node enters the scene tree for the first time.
func _ready() -> void:
	# ƒë·∫£m b·∫£o signal ƒë∆∞·ª£c connect
	body_entered.connect(_on_body_entered)
	pass # Replace with function body.

# Called every frame. 'delta' is the elapsed time since the previous frame.
func _process(delta: float) -> void:
	global_position += direction * speed * delta
	pass

func _on_body_entered(body: Node):
	if body.is_in_group("enemy"):
		print_debug("body.hit()")
		if body.has_method("hit"):
			body.hit(damage)
			queue_free()   # üí• hu·ª∑ ƒë·∫°n
			spawn_explosion()

func spawn_explosion() -> void:
	# T·∫°o explosion
	var explosion = EXPLOSION.instantiate()
	
	# ƒê·∫∑t v·ªã tr√≠ = v·ªã tr√≠ ƒë·∫°n
	explosion.global_position = global_position
	
	# C√≥ th·ªÉ random xoay cho ƒë·∫πp
	explosion.rotation = randf() * TAU
	
	# M√†u cho n·ªï l·ª≠a
	#explosion.modulate = Color(0.759, 0.322, 0.475, 1.0)
	
	# Random scale t·ª´ 3 ƒë·∫øn 4
	var random_scale = randf_range(3, 4)
	explosion.scale = Vector2(random_scale, random_scale)
	
	# Add v√†o scene ch√≠nh (kh√¥ng ph·∫£i v√†o bullet!)
	get_tree().current_scene.add_child(explosion)
	
	print("üí• Explosion created at: ", global_position)
